addons/payment/models/payment_acquirer.py

```diff
@@ -224,66 +197,65 @@ def render(self, cr, uid, id, reference, amount, currency_id, partner_id=False,
-            values['country'] = self.pool['res.country'].browse(cr, uid, values.get('partner_country_id'), context=context)
-
+            values['country'] = self.env['res.country'].browse(values['partner_country_id'])
 
         # compute fees
-        fees_method_name = '%s_compute_fees' % acquirer.provider
+        fees_method_name = '%s_compute_fees' % self.provider
         if hasattr(self, fees_method_name):
-            fees = getattr(self, fees_method_name)(cr, uid, id, values['amount'], values['currency_id'], values['partner_country_id'], context=None)
-            values['fees'] = float_round(fees, 2)
+            fees = getattr(self, fees_method_name)(values['amount'], values['currency_id'], values['partner_country_id'])
+            values['fees'] = tools.float_round(fees[0], 2)
 
         # call <name>_form_generate_values to update the tx dict with acqurier specific values
-        cust_method_name = '%s_form_generate_values' % (acquirer.provider)
+        cust_method_name = '%s_form_generate_values' % (self.provider)
         if hasattr(self, cust_method_name):
             method = getattr(self, cust_method_name)
-            values = method(cr, uid, id, values, context=context)
+            inv_type = method(values)
+            values = inv_type[0] if isinstance(inv_type, list) else inv_type
 
         values.update({
-            'tx_url': context.get('tx_url', self.get_form_action_url(cr, uid, id, context=context)),
-            'submit_class': context.get('submit_class', 'btn btn-link'),
-            'submit_txt': context.get('submit_txt'),
-            'acquirer': acquirer,
-            'user': self.pool.get("res.users").browse(cr, uid, uid, context=context),
-            'context': context,
+            'tx_url': self.env.context.get('tx_url', self.get_form_action_url()),
+            'submit_class': self.env.context.get('submit_class', 'btn btn-link'),
+            'submit_txt': self.env.context.get('submit_txt'),
+            'acquirer': self,
+            'user': self.env.user,
+            'context': self.env.context,
             'type': values.get('type') or 'form',
         })
         values.setdefault('return_url', False)
 
         # because render accepts view ids but not qweb -> need to use the xml_id
-        return self.pool['ir.ui.view'].render(cr, uid, acquirer.view_template_id.xml_id, values, engine='ir.qweb', context=context)
-
-    def _registration_render(self, cr, uid, id, partner_id, qweb_context=None, context=None):
-        acquirer = self.browse(cr, uid, id, context=context)
-        if qweb_context is None:
-            qweb_context = {}
-        qweb_context.update(id=id, partner_id=partner_id)
-        method_name = '_%s_registration_form_generate_values' % (acquirer.provider,)
+        return self.env['ir.ui.view']._model.render(self.env.cr, self.env.uid, self.view_template_id.xml_id, values, context=self.env.context)
+
+    def _registration_render(self, partner_id, values={}):
```

Comment:
where does this comes from? 
```diff
-    def _registration_render(self, cr, uid, id, partner_id, qweb_context=None, context=None):
+    def _registration_render(self, partner_id, values={}):
```
Moreover putting an empty dictionary in signature is a well known [python pitfall](http://www.andre-bubel.de/blog/2012/10/14/python-pitfall-with-mutable-default-parameters/)